src/storage/src/client.rs

```diff
@@ -0,0 +1,325 @@
+/// ```
+/// # tokio_test::block_on(async {
+/// # use google_cloud_storage::client::Storage;
+/// let client = Storage::builder().build().await?;
+/// // use `client` to make requests to Cloud Storage.
+/// # gax::Result::<()>::Ok(()) });
+/// ```
+///
+/// # Configuration
+///
+/// To configure `Storage` use the `with_*` methods in the type returned
+/// by [builder()][Storage::builder]. The default configuration should
+/// work for most applications. Common configuration changes include
+///
+/// * [with_endpoint()]: by default this client uses the global default endpoint
+///   (`https://storage.googleapis.com`). Applications using regional
+///   endpoints or running in restricted networks (e.g. a network configured
+//    with [Private Google Access with VPC Service Controls]) may want to
+///   override this default.
+/// * [with_credentials()]: by default this client uses
+///   [Application Default Credentials]. Applications using custom
+///   authentication may need to override this default.
+///
+/// # Pooling and Cloning
+///
+/// `Storage` holds a connection pool internally, it is advised to
+/// create one and the reuse it.  You do not need to wrap `Storage` in
+/// an [Rc](std::rc::Rc) or [Arc](std::sync::Arc) to reuse it, because it
+/// already uses an `Arc` internally.
+///
+/// # Service Description
+///
+/// The Cloud Storage API allows applications to read and write data through
+/// the abstractions of buckets and objects. For a description of these
+/// abstractions please see <https://cloud.google.com/storage/docs>.
+///
+/// This client is used to perform metadata operations, such as creating
+/// buckets, deleting objects, listing objects, etc. It does not expose any
+/// functions to write or read data in objects.
+///
+/// Resources are named as follows:
+///
+/// - Projects are referred to as they are defined by the Resource Manager API,
+///   using strings like `projects/123456` or `projects/my-string-id`.
+///
+/// - Buckets are named using string names of the form:
+///   `projects/{project}/buckets/{bucket}`
+///   For globally unique buckets, `_` may be substituted for the project.
+///
+/// - Objects are uniquely identified by their name along with the name of the
+///   bucket they belong to, as separate strings in this API. For example:
+///   ```no_rust
+///   bucket = "projects/_/buckets/my-bucket"
+///   object = "my-object/with/a/folder-like/name"
+///   ```
+///   Note that object names can contain `/` characters, which are treated as
+///   any other character (no special directory semantics).
+///
+/// [with_endpoint()]: ClientBuilder::with_endpoint
+/// [with_credentials()]: ClientBuilder::with_credentials
+/// [Private Google Access with VPC Service Controls]: https://cloud.google.com/vpc-service-controls/docs/private-connectivity
+/// [Application Default Credentials]: https://cloud.google.com/docs/authentication#adc
+#[derive(Clone, Debug)]
+pub struct Storage {
+    inner: reqwest::Client,
```

Comment:
why do we want to use a `reqwest::Client` directly, instead of our fancy schmancy wrapper?

Reply:
I imagine `read_object` is retry-able.

Reply:
> why do we want to use a reqwest::Client directly, instead of our fancy schmancy wrapper?

Because we will need to do all kinds of ad-hoc things:

- We will need to pool multiple connections.
- We will need to remove connections from the pool when a request fails with 429 or 503 (or anything that hints at an overloaded server or front-end).
- We will need to stream data, which our client cannot do and probably should not do either.

> I imagine read_object is retry-able.

Resumable.  You would need to start from the last byte received.  So I am not sure we would even use the same retry policies.

Reply:
Ack, thanks.